akka-typed/src/main/scala/akka/typed/ActorContext.scala

```diff
@@ -0,0 +1,238 @@
+import scala.reflect.classTag
+import java.util.concurrent.ConcurrentLinkedQueue
+import scala.annotation.tailrec
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.ExecutionContextExecutor
+import akka.typed.Behavior.stoppedBehavior
+
+/**
+ * An Actor is given by the combination of a [[Behavior]] and a context in
+ * which this behavior is executed. As per the Actor Model an Actor can perform
+ * the following actions when processing a message:
+ *
+ *  - send a finite number of messages to other Actors it knows
+ *  - create a finite number of Actors
+ *  - designate the behavior for the next message
+ *
+ * In Akka the first capability is accessed by using the `!` or `tell` method
+ * on an [[ActorRef]], the second is provided by [[ActorContext#spawn]]
+ * and the third is implicit in the signature of [[Behavior]] in that the next
+ * behavior is always returned from the message processing logic.
+ *
+ * An `ActorContext` in addition provides access to the Actor’s own identity (“`self`”),
+ * the [[ActorSystem]] it is part of, methods for querying the list of child Actors it
+ * created, access to [[Terminated DeathWatch]] and timed message scheduling.
+ */
+trait ActorContext[T] {
+
+  /**
+   * The identity of this Actor, bound to the lifecycle of this Actor instance.
+   * An Actor with the same name that lives before or after this instance will
+   * have a different [[ActorRef]].
+   */
+  def self: ActorRef[T]
+
+  /**
+   * The [[Props]] from which this Actor was created.
+   */
+  def props: Props[T]
+
+  /**
+   * The [[ActorSystem]] to which this Actor belongs.
+   */
+  def system: ActorSystem[Nothing]
+
+  /**
+   * The list of child Actors created by this Actor during its lifetime that
+   * are still alive, in no particular order.
+   */
+  def children: Iterable[ActorRef[Nothing]]
+
+  /**
+   * The named child Actor if it is alive.
+   */
+  def child(name: String): Option[ActorRef[Nothing]]
+
+  /**
+   * Create a child Actor from the given [[Props]] under a randomly chosen name.
+   * It is good practice to name Actors wherever practical.
+   */
+  def spawnAnonymous[U](props: Props[U]): ActorRef[U]
+
+  /**
+   * Create a child Actor from the given [[Props]] and with the given name.
+   */
+  def spawn[U](props: Props[U], name: String): ActorRef[U]
+
+  /**
+   * Create an untyped child Actor from the given [[akka.actor.Props]] under a randomly chosen name.
+   * It is good practice to name Actors wherever practical.
+   */
+  def actorOf(props: untyped.Props): untyped.ActorRef
+
+  /**
+   * Create an untyped child Actor from the given [[akka.actor.Props]] and with the given name.
+   */
+  def actorOf(props: untyped.Props, name: String): untyped.ActorRef
+
+  /**
+   * Force the child Actor under the given name to terminate after it finishes
+   * processing its current message. Nothing happens if no Actor exists with the
+   * given name.
+   */
+  def stop(childName: String): Unit
```

Comment:
Accepting an ActorRef does not convey the information that only direct child actors are permissible while giving a name does not verify the incarnation ID. What a pickle. We could of course completely refactor all of Akka to use path elements of type `(String, Int)` instead of just `String` and then `stop` would require such a path element. WDYT? I certainly need to think a little more about this.